reinforcement_learning/agent/rl_agent.py

```diff
@@ -0,0 +1,144 @@
+            self.data_config, self.isBot, self.new_user
+        )
+
+        if config.get("nlu_type", "") == "neural":
+            self.nlu = NeuralNLU(None)
+
+    def initialize(self) -> None:
+        """Initializes the agent."""
+        self.dialogue_manager.initialize()
+
+    def generate_utterance(
+        self,
+        agent_dacts: List[DialogueAct],
+        options: DialogueOptions = {},
+        user_fname: str = None,
+        recommended_item: Dict[str, Any] = None,
+    ):
+        """Generates an utterance object based on agent dialogue acts.
+
+        Args:
+            agent_dacts: Agent dialogue acts.
+            options: Dialogue options that are provided to the user along with
+              the utterance.
+            user_fname: User's first name. Defaults to None.
+            recommended_item: Recommended item. Defaults to None.
+
+        Returns:
+            An annotated utterance and the associated options.
+        """
+        metadata = {"options": options}
+        if recommended_item:
+            metadata.update({"recommended_item": recommended_item})
+
+        agent_response, options = self.nlg.generate_output(
+            agent_dacts,
+            self.dialogue_manager.get_state(),
+            user_fname=user_fname,
+        )
+        agent_intent = Intent(
+            ";".join([da.intent.value.label for da in agent_dacts])
+        )
+
+        if not self.isBot:
+            logger.debug(
+                str(self.dialogue_manager.dialogue_state_tracker.dialogue_state)
+            )
+
+            utterance = AnnotatedUtterance(
+                intent=agent_intent,
+                text=agent_response,
+                participant=DialogueParticipant.AGENT,
+                annotations=[],
+                metadata=metadata,
+            )
+        else:
+            record_data = self.dialogue_manager.get_state().to_dict()
+            metadata.update({"record_data": record_data})
+            utterance = AnnotatedUtterance(
+                intent=agent_intent,
+                text=agent_response,
+                participant=DialogueParticipant.AGENT,
+                annotations=[],
+                metadata=metadata,
+            )
+
+        return utterance, options
+
+    def get_user_dialogue_acts(
+        self,
+        user_utterance: UserUtterance,
+        utterance_options: DialogueOptions,
+    ) -> List[DialogueAct]:
+        """Generates dialogue acts associated to a given user utterance.
+
+        Args:
+            user_utterance: User utterance.
+            utterance_options: Dialogue options that are provided to the user
+              along with the utterance.
+
+        Returns:
+            List of dialogue acts.
+        """
+        return self.nlu.generate_dacts(
+            user_utterance,
+            utterance_options,
+            self.dialogue_manager.get_state(),
+        )
+
+    def update_placeholder_dialogue_act(
```

Comment:
Nit: The purpose of this method is not clear. Maybe add some more information to the docstrings.

Reply:
Good point, I removed the method and updated the docstring of `get_filled_dialogue_acts`. Please check that the purpose is clear now.